Sources/Foundation/FileManager+Win32.swift

```diff
@@ -13,6 +13,96 @@
+    func withNormalizedRepresentation<Result>(of path: String, _ body: (UnsafePointer<WCHAR>) throws -> Result) rethrows -> Result {
+        var path = path
+
+        // Strip the leading `/` on a RFC8089 path (`/[drive-letter]:/...` ).  A
+        // leading slash indicates a rooted path on the drive for teh current
+        // working directory.
+        if path.count > 3 {
+            let index0 = path.startIndex
+            let index1 = path.index(path.startIndex, offsetBy: 1)
+            let index2 = path.index(path.startIndex, offsetBy: 2)
+
+            if path[index0] == "/", path[index1].isLetter, path[index2] == ":" {
+                path.removeFirst()
+            }
+        }
+
+        // Win32 APIs can support `/` for the arc separator. However,
+        // symlinks created with `/` do not resolve properly, so normalize
+        // the path.
+        path = path.replacing("/", with: "\\")
+
+        // Droop trailing slashes unless it follows a drive specification.  The
+        // trailing arc separator after a drive specifier iindicates the root as
+        // opposed to a drive relative path.
+        while path.count > 1, path[path.index(before: path.endIndex)] == "\\",
+                !(path.count == 3 &&
+                    path[path.index(path.endIndex, offsetBy: -2)] == ":" &&
+                    path[path.index(path.endIndex, offsetBy: -3)].isLetter) {
+            path.removeLast()
+        }
+
+        return try path.withCString(encodedAs: UTF16.self, body)
+    }
+
+    guard !path.isEmpty else {
+        throw NSError(domain: NSCocoaErrorDomain,
+                      code: CocoaError.fileReadInvalidFileName.rawValue,
+                      userInfo: [NSFilePathErrorKey:path])
+    }
+
+    return try withNormalizedRepresentation(of: path) { pwszPath in
+        guard !path.hasPrefix("\\\\"), !path.hasPrefix("//") else {
+            return try body(pwszPath)
+        }
+
+        let dwLength = GetFullPathNameW(pwszPath, 0, nil, nil)
+        let path = withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: Int(dwLength)) {
+            _ = GetFullPathNameW(pwszPath, DWORD($0.count), $0.baseAddress, nil)
+            return String(decodingCString: $0.baseAddress!, as: UTF16.self)
+        }
+        return try "\\\\?\\\(path)".withCString(encodedAs: UTF16.self, body)
+    }
+}
+
+private func walk(directory path: URL, _ body: (String, DWORD) throws -> Void) rethrows {
+    try "\(path.NTPath)\\*".withCString(encodedAs: UTF16.self) {
+        var ffd: WIN32_FIND_DATAW = .init()
+        let capacity = MemoryLayout.size(ofValue: ffd.cFileName) / MemoryLayout<WCHAR>.size
+
+        let hFind: HANDLE = FindFirstFileW($0, &ffd)
+        if hFind == INVALID_HANDLE_VALUE {
+            throw _NSErrorWithWindowsError(GetLastError(), reading: true, paths: [path.path])
+        }
+
+        defer { FindClose(hFind) }
+
+        repeat {
+            let entry: String = withUnsafePointer(to: ffd.cFileName) {
```

Comment:
Can you use with `withUnsafeBytes(of:)` here to avoid having to calculate the `capacity` value?

Reply:
I don't think that the string is guaranteed to be null terminated as it is a fixed size buffer.  The capacity is `sizeof(cFileName)`.

Reply:
That's my point - `withUnsafeBytes()` gives you an `UnsafeRawBufferPointer` of the appropriate size (which is going to be the size of the value you're accessing) instead of only an `UnsafePointer` - the former makes a lot more sense for accessing an imported C fixed-length array (which imports as a tuple, sigh).